cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +563,483 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
     return _partition_key_restrictions->bounds_ranges(options);
 }
 
+namespace {
+
+using namespace expr;
+
+clustering_key_prefix::prefix_equal_tri_compare get_unreversed_tri_compare(const schema& schema) {
+    clustering_key_prefix::prefix_equal_tri_compare cmp(schema);
+    std::vector<data_type> types = cmp.prefix_type->types();
+    for (auto& t : types) {
+        if (t->is_reversed()) {
+            t = t->underlying_type();
+        }
+    }
+    cmp.prefix_type = make_lw_shared<compound_type<allow_prefixes::yes>>(types);
+    return cmp;
+}
+
+/// True iff r1 start is strictly before r2 start.
+bool starts_before_start(
+        const query::clustering_range& r1,
+        const query::clustering_range& r2,
+        const clustering_key_prefix::prefix_equal_tri_compare& cmp) {
+    if (!r2.start()) {
+        return false; // r2 start is -inf, nothing is before that.
+    }
+    if (!r1.start()) {
+        return true; // r1 start is -inf, while r2 start is finite.
+    }
+    const auto diff = cmp(r1.start()->value(), r2.start()->value());
+    if (diff < 0) { // r1 start is strictly before r2 start.
+        return true;
+    }
+    if (diff > 0) { // r1 start is strictly after r2 start.
+        return false;
+    }
+    const auto len1 = r1.start()->value().representation().size();
```

Comment:
I suppose the way to generalize this is to parametrize prefix_equality_tri_comp with the roles of the two inputs: is it a start-bound or an end-bound? and to add inclusive/exclusive information to the bounds.